common/flatpak-transaction.c

```diff
@@ -3024,6 +3031,125 @@ flatpak_transaction_add_install_flatpakref (FlatpakTransaction *self,
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr(GPtrArray) install_refs = g_ptr_array_new_with_free_func (g_free);
+  g_autoptr(GPtrArray) preinstalled_refs = NULL;
+  g_auto(GStrv) remotes = NULL;
+  g_auto(GStrv) preinstall_paths = NULL;
+
+  remotes = flatpak_dir_list_remotes (priv->dir, NULL, error);
+  if (remotes == NULL)
+    return FALSE;
+
+  preinstalled_refs = flatpak_dir_get_config_patterns (priv->dir, "preinstalled");
+
+  preinstall_paths = flatpak_get_preinstall_config_file_paths (NULL, error);
+  if (preinstall_paths == NULL)
+    return FALSE;
+
+  /* Find preinstalls that should get installed */
+  for (int i = 0; preinstall_paths[i] != NULL; i++)
+    {
+      const char *path = preinstall_paths[i];
+      g_autoptr(FlatpakDecomposed) decomposed = NULL;
+      g_autofree char *collection_id = NULL;
+
+      if (!flatpak_parse_preinstall_config_file (path, priv->default_arch, &decomposed, &collection_id, error))
+        return FALSE;
+
+      g_info ("Found preinstall ref %s from config file %s", flatpak_decomposed_get_ref (decomposed), path);
+
+      /* Store for later */
+      g_ptr_array_add (install_refs, flatpak_decomposed_dup_ref (decomposed));
+
+      /* Skip over if it's listed as previously preinstalled - it's now under
+       * user's control and we no longer install it again, even if the user
+       * manually removes it. */
+      if (!priv->reinstall &&
+          flatpak_g_ptr_array_contains_string (preinstalled_refs, flatpak_decomposed_get_ref (decomposed)))
+        {
+          g_info ("Preinstall ref %s is marked as already preinstalled; skipping", flatpak_decomposed_get_ref (decomposed));
+          continue;
+        }
+
+      for (int j = 0; remotes[j] != NULL; j++)
+        {
+          const char *remote = remotes[j];
+          g_autoptr(GError) local_error = NULL;
+          g_autofree char *remote_collection_id = NULL;
+
+          if (flatpak_dir_get_remote_disabled (priv->dir, remote))
+            continue;
+
+          remote_collection_id = flatpak_dir_get_remote_collection_id (priv->dir, remote);
+
+          /* Choose the first match if the collection ID was not specified */
+          if (collection_id != NULL &&
+              g_strcmp0 (remote_collection_id, collection_id) != 0)
+            continue;
+
+          g_info ("Adding preinstall of %s from remote %s", flatpak_decomposed_get_ref (decomposed), remote);
+
+          if (!flatpak_transaction_add_ref (self, remote, decomposed, NULL, NULL, NULL,
+                                            FLATPAK_TRANSACTION_OPERATION_INSTALL,
+                                            NULL, NULL, FALSE, TRUE, NULL,
+                                            &local_error))
+            g_info ("Failed to add preinstall ref %s: %s", flatpak_decomposed_get_ref (decomposed),
+                     local_error->message);
+        }
```

Comment:
Isn't this missing a break?

Reply:
No, it's deliberately so and is meant to continue with other preinstall refs even if one fails and do as much as possible - this way, if the repo has issues or has dropped an app, it doesn't stop other apps from getting preinstalled. Compare this to auto-install ref handling in the same .c file, which has basically the same logic and is also similarly supposed to work headless.